common/src/test/java/tachyon/security/authentication/AuthenticationFactoryTest.java

```diff
@@ -22,49 +41,218 @@
+
+    // start server
+    startServerThread(mTachyonConf);
+
+    // when connecting, authentication happens. It is a no-op in Simple mode.
+    TTransport client = createClient("anyone");
+    client.open();
+    Assert.assertTrue(client.isOpen());
+
+    // clean up
+    client.close();
+    mServer.stop();
+  }
+
+  /**
+   * In CUSTOM mode, the TTransport mechanism is PLAIN. When server authenticate the connected
+   * client user, it use configured AuthenticationProvider.
+   */
+  @Test
+  public void customAuthenticationTest() throws Exception {
+    mTachyonConf.set(Constants.TACHYON_SECURITY_AUTHENTICATION, "CUSTOM");
+    mTachyonConf.set(Constants.TACHYON_AUTHENTICATION_PROVIDER_CUSTOM_CLASS,
+        PlainServerCallbackHandlerTest.NameMatchAuthenticationProvider.class.getName());
+
+    // start server
+    startServerThread(mTachyonConf);
+
+    // when connecting, authentication happens. Users starting with "tachyon" can pass.
+    TTransport client = createClient("tachyon");
+    client.open();
+    Assert.assertTrue(client.isOpen());
+
+    // Users not starting with "tachyon" can not pass, and throw exception.
+    TTransport wrongClient = createClient("not-tachyon");
+    mThrown.expect(TTransportException.class);
+    mThrown.expectMessage("Peer indicated failure: AuthenticationProvider authenticate failed: "
+        + "Only allow the user starting with tachyon");
+    try {
+      wrongClient.open();
+    } catch (TTransportException e) {
+      // clean up and re-throw the expected exception
+      client.close();
+      mServer.stop();
+      throw e;
+    }
+  }
+
+  /**
+   * TODO: In KERBEROS mode, ...
+   */
+  @Test
+  public void kerberosAuthenticationTest() throws Exception {
+    mTachyonConf.set(Constants.TACHYON_SECURITY_AUTHENTICATION, "KERBEROS");
+
+    // throw unsupported exception currently
+    mThrown.expect(UnsupportedOperationException.class);
+    mThrown.expectMessage("Kerberos is not supported currently.");
+    startServerThread(mTachyonConf);
+  }
+
+  private void startServerThread(TachyonConf conf) throws Exception {
+    // create args and use them to build a Thrift TServer
+    AuthenticationFactory factory = new AuthenticationFactory(conf);
+    TTransportFactory tTransportFactory = factory.getServerTransportFactory();
+
+    TServerSocket wrappedServerSocket = new TServerSocket(mServerAddress);
+
+    mServer = new TThreadPoolServer(new TThreadPoolServer.Args(wrappedServerSocket)
+        .maxWorkerThreads(2).minWorkerThreads(1)
+        .processor(null).transportFactory(tTransportFactory)
+        .protocolFactory(new TBinaryProtocol.Factory(true, true)));
+
+    // start the server in a new thread
+    Thread serverThread = new Thread(new Runnable() {
+      @Override
+      public void run() {
+        mServer.serve();
+      }
+    });
+
+    serverThread.start();
+
+    // ensure server is running
+    while (!mServer.isServing() && serverThread.isAlive()) {
+      Thread.sleep(50);
```

Comment:
Break if server doesn't start serving for some time, so to prevent infinite waiting on test results?